cache/s3cache/s3cache.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	//	setup the s3 session.
+	//	if the accessKey and secreteKey are not provided (static creds) then the provider chain is used
+	//	http://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html
+	s3cache.Client = s3.New(
+		session.New(&awsConfig),
+	)
+
+	//	in order to confirm we have the correct permissions on the bucket create a small file
+	//	and test a PUT, GET and DELETE to the bucket
+	key := cache.Key{
+		MapName:   "tegola-test-map",
+		LayerName: "test-layer",
+		Z:         0,
+		X:         0,
+		Y:         0,
+	}
+	//	write a test file
+	if err := s3cache.Set(&key, []byte("\x53\x69\x6c\x61\x73")); err != nil {
+		return nil, err
+	}
+
+	//	read the test file
+	_, hit, err := s3cache.Get(&key)
+	if err != nil {
+		return nil, err
+	}
+	if !hit {
+		//	return an error?
+	}
+
+	//	purge the test file
+	if err := s3cache.Purge(&key); err != nil {
+		return nil, err
+	}
+
+	return &s3cache, nil
+}
+
+type S3Cache struct {
+	//	Bucket is the name of the s3 bucket to operate on
+	Bucket string
+
+	//	MaxZoom determins the max zoom the cache to persist. Beyond this
+	//	zoom, cache Set() calls will be ignored. This is useful if the cache
+	//	should not be leveraged for higher zooms when data changes often.
+	MaxZoom *uint
+
+	//	client holds a reference to the s3 client. it's expected the client
+	//	has an active session and read, write, delete permissions have been checked
+	Client *s3.S3
+}
+
+func (s3c *S3Cache) Set(key *cache.Key, val []byte) error {
+	var err error
+
+	//	check for maxzoom
+	if s3c.MaxZoom != nil && key.Z <= int(*s3c.MaxZoom) {
+		return nil
+	}
+
+	input := s3.PutObjectInput{
+		Body:   aws.ReadSeekCloser(bytes.NewReader(val)),
+		Bucket: aws.String(s3c.Bucket),
+		Key:    aws.String(key.String()),
+	}
+
+	_, err = s3c.Client.PutObject(&input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			default:
+				return aerr
+			}
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (s3c *S3Cache) Get(key *cache.Key) ([]byte, bool, error) {
+	var err error
+
+	input := s3.GetObjectInput{
+		Bucket: aws.String(s3c.Bucket),
+		Key:    aws.String(key.String()),
+	}
+
+	result, err := s3c.Client.GetObject(&input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchKey:
+				return nil, false, nil
+			default:
+				return nil, false, aerr
+			}
+		}
+		return nil, false, err
+	}
+
+	var buf bytes.Buffer
+	_, err = io.Copy(&buf, result.Body)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return buf.Bytes(), true, nil
+}
+
+func (s3c *S3Cache) Purge(key *cache.Key) error {
+	var err error
+
+	input := s3.DeleteObjectInput{
+		Bucket: aws.String(s3c.Bucket),
+		Key:    aws.String(key.String()),
+	}
+
+	_, err = s3c.Client.DeleteObject(&input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			default:
+				return err
+			}
+		}
+		return err
+	}
+
+	return nil
```

Comment:
No need for the complicated error handling here. You are not doing anything with the error; so just return it.
```
return err
```